crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+            dirfd,
+            path,
+            buf,
+            buf_len,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_READLINK,
+            types::Rights::empty(),
+            types::Lookupflags::empty(),
+            path,
+            false,
+        )?;
+
+        // This is to account for `path_readlink` call with zero-length
+        // buffer. This is a valid POSIX call hence we need to accommodate.
+        let slice = if buf_len > 0 {
+            let as_arr = buf.as_array(buf_len);
+            let as_raw = as_arr.as_raw(&mut bc)?;
+            unsafe { &mut *as_raw }
+        } else {
+            &mut []
+        };
+
+        let host_bufused = match resolved.dirfd() {
+            Descriptor::VirtualFile(_virt) => {
+                unimplemented!("virtual readlink");
+            }
+            _ => path::readlink(resolved, slice)?,
+        };
+        let host_bufused = host_bufused.try_into()?;
+
+        trace!("     | (buf_ptr,*buf_used)={:?}", slice);
+        trace!("     | *buf_used={:?}", host_bufused);
+
+        Ok(host_bufused)
+    }
+
+    fn path_remove_directory(&self, dirfd: types::Fd, path: &GuestPtr<'_, str>) -> Result<()> {
+        trace!("path_remove_directory(dirfd={:?}, path={:?})", dirfd, path);
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_REMOVE_DIRECTORY,
+            types::Rights::empty(),
+            types::Lookupflags::empty(),
+            path,
+            true,
+        )?;
+
+        debug!("path_remove_directory resolved={:?}", resolved);
+
+        match resolved.dirfd() {
+            Descriptor::VirtualFile(virt) => virt.remove_directory(resolved.path()),
+            _ => path::remove_directory(resolved),
+        }
+    }
+
+    fn path_rename(
+        &self,
+        old_fd: types::Fd,
+        old_path: &GuestPtr<'_, str>,
+        new_fd: types::Fd,
+        new_path: &GuestPtr<'_, str>,
+    ) -> Result<()> {
+        trace!(
+            "path_rename(old_fd={:?}, old_path={:?}, new_fd={:?}, new_path={:?})",
+            old_fd,
+            old_path,
+            new_fd,
+            new_path,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let old_as_str = old_path.as_raw(&mut bc)?;
+        let old_path = unsafe { &*old_as_str };
+        let new_as_str = new_path.as_raw(&mut bc)?;
+        let new_path = unsafe { &*new_as_str };
```

Comment:
Technically the FIXME items elsewhere are equally applicable here, there's not really any fundamental reason why these two guest buffers shouldn't be allowed to overlap.